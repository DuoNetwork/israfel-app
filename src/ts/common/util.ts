import { Constants, IUserOrder, Util as CommonUtil } from '@finbook/israfel-common';
import * as d3 from 'd3';
import moment from 'moment';
import * as CST from './constants';
import { ICustodianInfo } from './types';

class Util {
	public convertUpdateTime(timestamp: number): string {
		const diff = CommonUtil.getUTCNowTimestamp() - timestamp;
		if (diff < 60000) return 'just now';
		else if (diff < 3600000) return Math.floor(diff / 60000) + ' min ago';
		else if (diff < 86400000) return Math.floor(diff / 3600000) + ' hrs ago';
		else return 'long time ago';
	}

	public formatExpiry(isWeek: boolean) {
		return moment(CommonUtil.getExpiryTimestamp(isWeek)).format('YYYY-MM-DD HH:mm');
	}

	public formatTime(time: any) {
		return moment(time).format('MM-DD HH:mm:ss');
	}

	public formatTimeSecond(time: number) {
		return moment(time).format('YYYY-MM-DD HH:mm:ss');
	}

	public formatPriceShort(num: number) {
		if (num < 1) return d3.format('.3f')(num);
		return d3
			.format('.4s')(num)
			.toUpperCase()
			.replace(/G/g, 'B');
	}

	public formatStrike(num: number) {
		return d3.format('.4f')(num);
	}

	public formatBalance(num: number) {
		if (Math.abs(num) < 1e-4) return '0.000';
		return d3
			.format(Math.abs(num) > 1 ? ',.4s' : ',.4n')(num)
			.toUpperCase()
			.replace(/G/g, 'B');
	}

	public formatPercent(num: number) {
		return d3.format('.0%')(num);
	}

	public formatPercentAcc(num: number) {
		return d3.format('.2%')(num);
	}

	public formatNumber(num: number) {
		if (Math.abs(num) < 1e-4) return '0.000';
		if (Math.abs(num) < 1) return d3.format(',.4n')(num);
		if (Math.abs(num) < 100000) return d3.format(',.2f')(num);
		return d3
			.format(',.4s')(num)
			.toUpperCase()
			.replace(/G/g, 'B');
	}

	public formatMaturity(maturity: number) {
		return maturity ? moment(maturity).format('YYYY-MM-DD HH:mm') : CST.TH_PERPETUAL;
	}

	public getOrderTitle(order: IUserOrder) {
		if (order.type === Constants.DB_ADD) return 'Order Placed';
		else if (order.type === Constants.DB_TERMINATE)
			if (order.status === Constants.DB_FILL) return 'Order Fully Filled';
			else if (order.status === Constants.DB_TERMINATE) return 'Order Expired';
			// order.status === Constants.DB_CONFIRMED
			else return 'Order Cancelled';
		else if (order.status === Constants.DB_MATCHING) return 'Order is Being Filled';
		// if (order.status === Constants.DB_PFILL)
		else return 'Order Partially Filled';
	}

	public getOrderFullDescription(order: IUserOrder) {
		const [code1, code2] = order.pair.split('|');
		let baseDescription = `${order.side === Constants.DB_BID ? CST.TH_BUY : CST.TH_SELL} ${
			order.amount
		} ${code1} at ${order.price} ${code2} per ${code1}.`;
		if (order.type === Constants.DB_ADD)
			return baseDescription + ` Valid till ${this.formatMaturity(order.expiry)}`;

		if (order.type === Constants.DB_TERMINATE && order.status === Constants.DB_FILL)
			return baseDescription + ' Fully filled';

		if (order.fill || order.matching) {
			const totalFill =
				order.fill + (order.type === Constants.DB_TERMINATE ? 0 : order.matching);
			baseDescription += ` ${totalFill}(${this.formatPercent(
				totalFill / order.amount
			)}) filled.`;
		}

		if (order.matching && order.type !== Constants.DB_TERMINATE)
			baseDescription += 'Pending settlement';

		if (order.type === Constants.DB_TERMINATE && order.status === Constants.DB_CONFIRMED)
			return baseDescription + ' Cancelled by user.';

		if (order.type === Constants.DB_TERMINATE && order.status === Constants.DB_TERMINATE)
			return baseDescription + ' Expired.';

		if (order.type === Constants.DB_TERMINATE && order.status === Constants.DB_BALANCE)
			return baseDescription + ` Cancelled due to insufficient balance.`;

		if (order.type === Constants.DB_TERMINATE && order.status === Constants.DB_MATCHING)
			return baseDescription + ` Cancelled due to settlement error.`;

		if (order.type === Constants.DB_TERMINATE && order.status === Constants.DB_RESET)
			return baseDescription + ` Cancelled due to custodian reset.`;

		return baseDescription;
	}

	public getOrderDescription(order: IUserOrder) {
		const [code1, code2] = order.pair.split('|');
		return `${order.side === Constants.DB_BID ? CST.TH_BUY : CST.TH_SELL} ${
			order.amount
		} ${code1} at ${order.price} ${code2} per ${code1}. Total ${util.formatBalance(
			order.fill + (order.type === Constants.DB_TERMINATE ? 0 : order.matching)
		)} filled.`;
	}

	public getVersionDescription(order: IUserOrder) {
		if (order.type === Constants.DB_ADD) return 'Order placed.';

		if (order.type === Constants.DB_UPDATE) {
			let description = `Total ${this.formatBalance(order.fill + order.matching)} filled`;
			if (order.matching) description += ', pending settlement';
			return description;
		}

		if (order.type === Constants.DB_TERMINATE)
			if (order.status === Constants.DB_TERMINATE) return 'Expired';
			else if (order.status === Constants.DB_BALANCE)
				return 'Cancelled due to insufficent balance';
			else if (order.status === Constants.DB_MATCHING)
				return 'Cancelled due to settlement error';
			else if (order.status === Constants.DB_RESET) return 'Cancelled due to custodian reset';
			else if (order.status === Constants.DB_FILL) return 'Fully filled';
			// (order.status === Constants.DB_CONFIRMED)
			else return 'Cancelled by user';
		return 'Invalid order';
	}

	public getMaturityDescription(info?: ICustodianInfo) {
		return `This contract ${
			info && info.code.endsWith('PPT')
				? 'never expires'
				: 'expires in ' +
				(info
					? Math.floor(
							(info.states.maturity - CommonUtil.getUTCNowTimestamp()) / 86400000
						).toFixed(0)
					: 0) +
				' days'
		}.`;
	}

	public convertOrdersToCSV(orders: { [pair: string]: IUserOrder[] }): string {
		const headers = [
			CST.TH_PAIR,
			CST.TH_ORDER_HASH,
			CST.TH_VERSION,
			CST.TH_TIME_UTC,
			CST.TH_SIDE,
			CST.TH_PRICE,
			CST.TH_AMOUNT,
			CST.TH_FILL,
			CST.TH_FEE,
			CST.TH_EXPIRY_UTC,
			CST.TH_STATUS,
			CST.TH_TX_HASH
		];
		let output = '';
		headers.forEach(
			(header, index) =>
				(output += this.wrapCSVString(header) + (index < headers.length - 1 ? ',' : '\n'))
		);
		const ordersByHash: { [orderHash: string]: IUserOrder[] } = {};
		for (const pair in orders)
			orders[pair].forEach(order => {
				if (!ordersByHash[order.orderHash]) ordersByHash[order.orderHash] = [];
				ordersByHash[order.orderHash].push(order);
			});

		const rows: any[][] = [];
		for (const orderHash in ordersByHash) {
			ordersByHash[orderHash].sort((a, b) => a.currentSequence - b.currentSequence);
			ordersByHash[orderHash].forEach((order, index) => {
				const row: any[] = [];
				row.push(order.createdAt);
				row.push(order.pair.replace('|', '-')); // CST.TH_PAIR,
				row.push(order.orderHash); // CST.TH_ID,
				row.push(index + 1); // CST.TH_VERSION,
				row.push(
					moment.utc(order.updatedAt || order.createdAt).format('YYYY-MM-DD HH:mm:ss')
				); // CST.TH_TIME_UTC,
				row.push(order.side === Constants.DB_BID ? CST.TH_BUY : CST.TH_SELL); // CST.TH_SIDE,
				row.push(order.price); // CST.TH_PRICE,
				row.push(order.amount); // CST.TH_AMOUNT,
				row.push(order.fill + order.matching); // CST.TH_FILL,
				row.push(order.fee + ' ' + order.feeAsset); // CST.TH_FEE,
				row.push(moment.utc(order.expiry).format('YYYY-MM-DD HH:mm:ss')); // CST.TH_EXPIRY,
				// CST.TH_STATUS,
				if (order.type === Constants.DB_TERMINATE)
					if (order.status === Constants.DB_BALANCE)
						row.push('Cancelled due to insufficent balance');
					else if (order.status === Constants.DB_MATCHING)
						row.push('Cancelled due to settlement error');
					else if (order.status === Constants.DB_RESET)
						row.push('Cancelled due to custodian reset');
					else if (order.status === Constants.DB_FILL) row.push('Fully filled');
					else if (order.status === Constants.DB_CONFIRMED) row.push('Cancelled by user');
					else row.push('Expired');
				else if (order.status === Constants.DB_MATCHING) row.push('Pending settlement');
				else row.push('Open');
				row.push(order.transactionHash || ''); // CST.TH_TX_HASH
				rows.push(row);
			});
		}

		rows.sort((a, b) => a[1].localeCompare(b[1]) || -a[0] + b[0] || -a[3] + b[3]);
		rows.forEach(row =>
			row.forEach((item, index) => {
				if (index)
					output +=
						this.wrapCSVString(item + '') + (index !== row.length - 1 ? ',' : '\n');
			})
		);

		return output;
	}

	public wrapCSVString(input: string): string {
		return ('' + input).indexOf(',') >= 0 ? '"' + input + '"' : input;
	}

	private getEtherScanLink() {
		return `https://${__ENV__ === Constants.DB_LIVE ? '' : 'kovan.'}etherscan.io`;
	}

	public getEtherScanTransactionLink(txHash: string) {
		return `${this.getEtherScanLink()}/tx/${txHash}`;
	}

	public getEtherScanTokenLink(tokenAddr: string, account: string = '') {
		return `${this.getEtherScanLink()}/token/${tokenAddr}${account ? `?a=${account}` : ''}`;
	}

	public getEtherScanAddressLink(account: string) {
		return `${this.getEtherScanLink()}/address/${account}`;
	}

	public fillZero(num: number) {
		return num < 10 ? `0${num}` : `${num}`;
	}

	public isEmptyObject(obj: object | undefined | null): boolean {
		if (!obj) return true;

		for (const prop in obj) if (obj.hasOwnProperty(prop)) return false;

		return true;
	}
}

const util = new Util();
export default util;
