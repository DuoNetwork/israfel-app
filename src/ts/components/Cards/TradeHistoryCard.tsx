import * as React from 'react';
import * as CST from 'ts/common/constants';
import { INotification, ITrade } from 'ts/common/types';
import util from 'ts/common/util';
import { SDivFlexCenter } from '../_styled';
import { SCard, SCardList, SCardTitle } from './_styled';

interface IProps {
	trades: { [pair: string]: ITrade[] };
	notify: (notification: INotification) => any;
}

interface IState {
	expandIndex: number;
}

export default class TradeHistoryCard extends React.Component<IProps, IState> {
	constructor(props: IProps) {
		super(props);
		this.state = { expandIndex: 0 };
	}

	private handleShow(i: number) {
		this.setState({
			expandIndex: i === this.state.expandIndex ? -1 : i
		});
	}

	public render() {
		const { trades } = this.props;
		const { expandIndex } = this.state;

		const tradeList: ITrade[][] = [];
		const tradeKeylist: string[] = [];

		for (const key in trades) {
			const subTradeList: ITrade[] = [];
			for (const subKey in trades[key]) subTradeList.push(trades[key][subKey]);
			tradeKeylist.push(key);
			tradeList.push(subTradeList);
		}
		const showTradeList: object[] = [];
		tradeList.forEach((c, i) => {
			const showSubTradeList: object[] = [];
			c.forEach((d, j) => {
				showSubTradeList.push(
					<li key={j} style={{ padding: '5px 5px' }}>
						<span className="title">{`Px:${util.formatPriceShort(
							d.taker.price
						)}`}</span>
						<span className="content">{`Amt:${util.formatNumber(
							d.taker.amount
						)}`}</span>
						<span className="content">{`Time:${util.formatTime(d.timestamp)}`}</span>
						<span className="content">{`Side:${d.taker.side}`}</span>
					</li>
				);
			});
			showTradeList.push(
				<li key={i} style={{ padding: '5px 5px', flexDirection: 'column' }}>
					<div
						style={{
							display: 'flex',
							justifyContent: 'space-between',
							cursor: 'pointer'
						}}
						onClick={() => this.handleShow(i)}
					>
						<span className="title">{tradeKeylist[i]}</span>
						<span className="content">{i === expandIndex ? '-' : '+'}</span>
					</div>
					<ul
						style={{
							paddingLeft: 10,
							maxHeight: i === expandIndex ? 200 : 0,
							overflow: 'hidden'
						}}
					>
						{showSubTradeList.length ? showSubTradeList : 'No Trades'}
					</ul>
				</li>
			);
		});

		return (
			<SCard
				title={<SCardTitle>{CST.TH_MARKET + ' ' + CST.TH_TRADES}</SCardTitle>}
				width="740px"
				margin="0 10px 20px 10px"
			>
				<SDivFlexCenter horizontal style={{ marginTop: '5px' }}>
					<SCardList noMargin>
						<div className="status-list-wrapper">
							<ul>{showTradeList.length ? showTradeList : '-'}</ul>
						</div>
					</SCardList>
				</SDivFlexCenter>
			</SCard>
		);
	}
}
