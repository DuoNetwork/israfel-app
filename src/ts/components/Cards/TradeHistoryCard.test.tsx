/**
 * @jest-environment jsdom
 */
import { shallow } from 'enzyme';
import * as React from 'react';
// import { SButton, SInput, SSlider } from './_styled';
import TradeHistoryCard from './TradeHistoryCard';
import util from 'ts/common/util';

describe('TradeHistoryCard Test', () => {
	describe('Test Snapshot', () => {
		it('Test Snapshot', () => {
			util.formatTime = jest.fn(() => '01-01 00:00:00');

			const wrapper1 = shallow(
				<TradeHistoryCard
					trades={{
						'aETH|WETH': [
							{
								pair: 'test',
								transactionHash: 'txHash',
								taker: {
									orderHash: 'test',
									address: 'test',
									side: 'test',
									price: 123,
									amount: 123,
									fee: 123
								},
								maker: { orderHash: 'test', price: 123, amount: 123, fee: 123 },
								feeAsset: 'test',
								timestamp: 123
							}
						]
					}}
					tokens={[
						{
							custodian: 'test',
							address: 'test',
							code: 'aETH',
							denomination: 123,
							precisions: { WETH: 123 },
							feeSchedules: { WETH: { rate: 123, minimum: 123 } },
							maturity: 123
						}
					]}
					tokenBalances={[
						{
							code: 'aETH',
							balance: 123,
							address: '0x00be45fe5903ab1b33a9d3969b05b29552a6d18b'
						}
					]}
				/>
			);
			expect(wrapper1).toMatchSnapshot();
		});

		it('Test getDerivedStateFromProps', () => {
			util.formatTime = jest.fn(() => '01-01 00:00:00');

			const wrapper1 = shallow(
				<TradeHistoryCard
					trades={{
						'MOZART-PPT|WETH': [
							{
								pair: 'test',
								transactionHash: 'test',
								taker: {
									orderHash: 'test',
									address: 'test',
									side: 'bid',
									price: 123,
									amount: 123,
									fee: 123
								},
								maker: { orderHash: 'test', price: 12, amount: 123, fee: 123 },
								feeAsset: 'test',
								timestamp: 123
							}
						]
					}}
					tokens={[
						{
							custodian: 'test',
							address: '0x00be45fe5903ab1b33a9d3969b05b29552a6d18b',
							code: 'MOZART-PPT',
							denomination: 12,
							precisions: { test: 123 },
							feeSchedules: { test: { asset: 'test', rate: 123, minimum: 123 } },
							maturity: 123
						}
					]}
					tokenBalances={[
						{
							code: 'MOZART-PPT',
							balance: 123,
							address: '0x00be45fe5903ab1b33a9d3969b05b29552a6d18b'
						}
					]}
				/>
			);
			wrapper1.setState({ pair: '' });
			wrapper1.setProps({
				tokenBalances: [
					{
						code: 'MOZART-PPT',
						balance: 123,
						address: '0x00be45fe5903ab1b33a9d3969b05b29552a6d18b'
					}
				]
			});
			expect(wrapper1).toMatchSnapshot();
			expect(wrapper1).toMatchSnapshot();
		});
	});
});
