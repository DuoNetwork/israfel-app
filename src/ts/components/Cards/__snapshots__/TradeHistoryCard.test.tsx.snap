// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TradeHistoryCard Test Test Snapshot Test Snapshot 1`] = `
<Styled(Card)
  extra={
    <Styled(Select)
      onSelect={[Function]}
      style={
        Object {
          "width": 200,
        }
      }
      value="aETH|WETH"
    >
      <Option
        className="optionMarketTrades"
        value="aETH|WETH"
      >
        aETH-WETH
      </Option>
    </Styled(Select)>
  }
  margin="0 10px 20px 10px"
  title={
    <styled.div>
      Market Trades
    </styled.div>
  }
  width="740px"
>
  <styled.div
    horizontal={true}
    style={
      Object {
        "marginTop": "5px",
      }
    }
  >
    <styled.div
      noMargin={true}
    >
      <div
        className="status-list-wrapper"
      >
        <Table
          bordered={false}
          className=""
          dataSource={
            Array [
              Object {
                "Amount": "123",
                "Link": "https://kovan.etherscan.io/tx/txHash",
                "Price": "123",
                "Side": "Sell",
                "Time": "01-01 00:00:00",
                "Type": "Limit",
                "key": "txHash",
              },
            ]
          }
          indentSize={20}
          loading={false}
          locale={Object {}}
          pagination={false}
          rowKey="key"
          showHeader={true}
          size="default"
          sortDirections={
            Array [
              "ascend",
              "descend",
            ]
          }
          style={
            Object {
              "border": "none",
              "width": "100%",
            }
          }
          useFixedHeader={false}
        >
          <Column
            className="columnAlignLeft"
            dataIndex="Time"
            title="Time"
          />
          <Column
            className="columnAlignLeft"
            dataIndex="Side"
            title="Side"
          />
          <Column
            className="columnAlignLeft"
            dataIndex="Type"
            title="Type"
          />
          <Column
            className="antdColumnAlignRight"
            dataIndex="Price"
            render={[Function]}
            title="Price"
          />
          <Column
            className="antdColumnAlignRight"
            dataIndex="Amount"
            title="Amount"
          />
          <Column
            className="imgColumn"
            dataIndex="Link"
            render={[Function]}
            title=""
          />
        </Table>
      </div>
    </styled.div>
  </styled.div>
</Styled(Card)>
`;

exports[`TradeHistoryCard Test Test Snapshot Test getDerivedStateFromProps 1`] = `
<Styled(Card)
  extra={
    <Styled(Select)
      onSelect={[Function]}
      style={
        Object {
          "width": 200,
        }
      }
      value="MOZART-PPT|WETH"
    >
      <Option
        className="optionMarketTrades"
        value="MOZART-PPT|WETH"
      >
        MOZART-PPT-WETH
      </Option>
    </Styled(Select)>
  }
  margin="0 10px 20px 10px"
  title={
    <styled.div>
      Market Trades
    </styled.div>
  }
  width="740px"
>
  <styled.div
    horizontal={true}
    style={
      Object {
        "marginTop": "5px",
      }
    }
  >
    <styled.div
      noMargin={true}
    >
      <div
        className="status-list-wrapper"
      >
        <Table
          bordered={false}
          className=""
          dataSource={
            Array [
              Object {
                "Amount": "120",
                "Link": "https://kovan.etherscan.io/tx/test",
                "Price": "12",
                "Side": "Buy",
                "Time": "01-01 00:00:00",
                "Type": "Market",
                "key": "test",
              },
            ]
          }
          indentSize={20}
          loading={false}
          locale={Object {}}
          pagination={false}
          rowKey="key"
          showHeader={true}
          size="default"
          sortDirections={
            Array [
              "ascend",
              "descend",
            ]
          }
          style={
            Object {
              "border": "none",
              "width": "100%",
            }
          }
          useFixedHeader={false}
        >
          <Column
            className="columnAlignLeft"
            dataIndex="Time"
            title="Time"
          />
          <Column
            className="columnAlignLeft"
            dataIndex="Side"
            title="Side"
          />
          <Column
            className="columnAlignLeft"
            dataIndex="Type"
            title="Type"
          />
          <Column
            className="antdColumnAlignRight"
            dataIndex="Price"
            render={[Function]}
            title="Price"
          />
          <Column
            className="antdColumnAlignRight"
            dataIndex="Amount"
            title="Amount"
          />
          <Column
            className="imgColumn"
            dataIndex="Link"
            render={[Function]}
            title=""
          />
        </Table>
      </div>
    </styled.div>
  </styled.div>
</Styled(Card)>
`;

exports[`TradeHistoryCard Test Test Snapshot Test getDerivedStateFromProps 2`] = `
<Styled(Card)
  extra={
    <Styled(Select)
      onSelect={[Function]}
      style={
        Object {
          "width": 200,
        }
      }
      value="MOZART-PPT|WETH"
    >
      <Option
        className="optionMarketTrades"
        value="MOZART-PPT|WETH"
      >
        MOZART-PPT-WETH
      </Option>
    </Styled(Select)>
  }
  margin="0 10px 20px 10px"
  title={
    <styled.div>
      Market Trades
    </styled.div>
  }
  width="740px"
>
  <styled.div
    horizontal={true}
    style={
      Object {
        "marginTop": "5px",
      }
    }
  >
    <styled.div
      noMargin={true}
    >
      <div
        className="status-list-wrapper"
      >
        <Table
          bordered={false}
          className=""
          dataSource={
            Array [
              Object {
                "Amount": "120",
                "Link": "https://kovan.etherscan.io/tx/test",
                "Price": "12",
                "Side": "Buy",
                "Time": "01-01 00:00:00",
                "Type": "Market",
                "key": "test",
              },
            ]
          }
          indentSize={20}
          loading={false}
          locale={Object {}}
          pagination={false}
          rowKey="key"
          showHeader={true}
          size="default"
          sortDirections={
            Array [
              "ascend",
              "descend",
            ]
          }
          style={
            Object {
              "border": "none",
              "width": "100%",
            }
          }
          useFixedHeader={false}
        >
          <Column
            className="columnAlignLeft"
            dataIndex="Time"
            title="Time"
          />
          <Column
            className="columnAlignLeft"
            dataIndex="Side"
            title="Side"
          />
          <Column
            className="columnAlignLeft"
            dataIndex="Type"
            title="Type"
          />
          <Column
            className="antdColumnAlignRight"
            dataIndex="Price"
            render={[Function]}
            title="Price"
          />
          <Column
            className="antdColumnAlignRight"
            dataIndex="Amount"
            title="Amount"
          />
          <Column
            className="imgColumn"
            dataIndex="Link"
            render={[Function]}
            title=""
          />
        </Table>
      </div>
    </styled.div>
  </styled.div>
</Styled(Card)>
`;
